### PR TITLE
feat: add `compose run` command

### LINKED ISSUE #270: Support Compose Run
First, this project is amazing!!!

Feature request. A lot of our Docker projects at Custom Ink leverage `docker-compose run` in various project bin files for local development. This allows us to use the compose file as a way to define a single place for environments, volumes, etc. Is `nerdctl compose run` anywhere in the future?

### PR DESCRIPTION
**Pull request by minuk-dev**
Signed-off-by: Min Uk Lee <minuk.dev@gmail.com>

close #270 

Changes (for more details, see each item below):
- remove `-d` from default RunArgs of serviceparser.Container
- add `apt install expect` into docker-compatibility test
- add Base.ComposeCmdWithHelper() into testutil
- add `compose run` command (no details below)
- add `compose run` docs into README.md (no details below)

Unimplmented Flags:
- `--no-TTY`, `--use-aliases`

Not tested Flags:
- volumes, entrypoint
 (I'll add additional test commit as soon as possible. I'm trying to figure out how to test it properly.)

### remove `-d` from default RunArgs of serviceparser.Container
- Previously, composer.upServiceContinaer() uses `-d` flag to find cid.

>   -d, --detach                      Run container in background and print container ID

- Because of it, cannot support `--interactive` flag of `compose run` and `StdinOpen` option of [compose-spec](https://github.com/compose-spec/compose-spec/blob/master/spec.md)
- Remove `-d` from default RunArgs and explicitly add property of container. Containers are now possible to determine whether or not to detach when they really start to run.

### add `apt install expect` into docker-compatibility test
- Currently, docker-compatibility test does not contain `unbuffer` command to emulate tty.
  - I found `unbuffer` command from [a test for run command](https://github.com/containerd/nerdctl/blob/c0c31637f3fbff37b2e55f13645465ecc078b646/cmd/nerdctl/run_linux_test.go#L181-L184).
- And even some tests are skipped because `unbuffer` is not installed.
  - [TestIssue108](https://github.com/containerd/nerdctl/runs/7995470842?check_suite_focus=true#step:5:354)
  - [TestExecTTY](https://github.com/containerd/nerdctl/runs/7995470842?check_suite_focus=true#step:5:795)
  - [TestRunTTY](https://github.com/containerd/nerdctl/runs/7995470842?check_suite_focus=true#step:5:1019)

### add Base.ComposeCmdWithHelper() into testutil
- As I mentioned, `unbuffer` needs to emulate tty to test `compose run`.
- Basically, I referenced Base.CmdWithHelper()

### CODE REVIEW

**Comment by minuk-dev** [comment]
I'm finding the reason why [docker compatibility test are failed](https://github.com/containerd/nerdctl/runs/8011633728?check_suite_focus=true#step:6:263)
As soon as possible, I'll fix it and push again.

**Comment by minuk-dev** [comment]
I added some test cases for `--entrypoint`, `--volume`.
Could anyone review, please?

File: pkg/composer/up_service.go [hunk 1]
```diff
@@ -146,8 +147,20 @@ func (c *Composer) upServiceContainer(ctx context.Context, service *serviceparse
 		logrus.Infof("Creating container %s", container.Name)
 	}
 
+	tempDir, err := os.MkdirTemp(os.TempDir(), "compose-")
+	if err != nil {
+		return "", fmt.Errorf("error while creating container %s: %w", container.Name, err)
```
Comment (fahedouch) [inline pkg/composer/up_service.go:152]: `error while creating/re-creating  container`

File: pkg/composer/up_service.go [hunk 2]
```diff
@@ -156,10 +169,29 @@ func (c *Composer) upServiceContainer(ctx context.Context, service *serviceparse
...
+
+	if service.Unparsed.StdinOpen {
+		cmd.Stdin = os.Stdin
+	}
+
+	if !container.Detached {
+		cmd.Stdout = os.Stdout
+	}
+
 	cmd.Stderr = os.Stderr
```
Comment (fahedouch) [inline pkg/composer/up_service.go:186]: same for `stderr` we should check the detach mode

File: pkg/composer/up_service.go [hunk 3]
```diff
@@ -146,8 +147,20 @@ func (c *Composer) upServiceContainer(ctx context.Context, service *serviceparse
...
+	defer os.RemoveAll(tempDir)
+	cidFilename := filepath.Join(tempDir, "cid")
+
+	if container.Detached && !service.Unparsed.StdinOpen && !service.Unparsed.Tty {
+		container.RunArgs = append([]string{"-d"}, container.RunArgs...)
+	}
+
 	//add metadata labels to container https://github.com/compose-spec/compose-spec/blob/master/spec.md#labels
 	container.RunArgs = append([]string{
+		"--cidfile=" + cidFilename,
```
[the remark this answers is not in the document]
Comment (minuk-dev) [reply pkg/composer/up_service.go:163]: I followed this comment.

https://github.com/containerd/nerdctl/blob/0ca024b1655a0d8505314d9d11ab2d75870d397f/pkg/composer/up_service.go#L129

Previously, we used `-d` flag at all times. So we can easily get container id through their stdout.
But, `-d` is optional now. So we should use `--cid-file`.

---
Here is the reason why upServiceContainer() should returns cid.

To control service's containers using their id. (e.g. stop containers)

https://github.com/containerd/nerdctl/blob/0ca024b1655a0d8505314d9d11ab2d75870d397f/pkg/composer/up_service.go#L58
https://github.com/containerd/nerdctl/blob/0ca024b1655a0d8505314d9d11ab2d75870d397f/pkg/composer/up_service.go#L63

**Review by AkihiroSuda — APPROVED** [review]
Thanks, and sorry for delay in reviewing

**Comment by AkihiroSuda** [comment]
Found a minor incompatibility: `docker compose run` randomize container name (`compose-wordpress_db_run_cca27158caa2 `) but `nerdctl compose run` does not (`compose-wordpress_db_1).

**Comment by fahedouch** [comment]
LGTM

**Comment by fahedouch** [comment]
@AkihiroSuda could you plz ask for squashing commits before merging the PR in the future :) !

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda could you plz ask for squashing commits before merging the PR in the future :) !

In this case having separate commits is fine, as each of the commits has a separate topic.

`Add tests and docs for compose run command ` could be meld into `Add compose run command template` though.
